Export products that are considered as Bundles

thisiscab opened the pull request:
This will export products that are considered as Bundles when using the following gem:
https://github.com/spree-contrib/spree-product-assembly

Shopify doesn't allow Bundles so we had to "hack" it. Which means that for each bundle of products on Solidus, we had to get all the possible parts (read variant) and permute all the possibilities. Those possibilities have to be uploaded on Shopify as Shopify variant.

Please see https://github.com/glossier/solidus_retail/issues/7 for more details

thisiscab committed:
WIP

thisiscab committed:
Separate updaters and exporters

The way that it works right now is that the exporter will export the
whole product, which means all of its variant.

Updaters will now only update self. So if we go in the admin and we
decide to update a unique product or a unique variant, then it will
update itself instead of saving the whole product will all of its
variant.

There is still some work to do in order to reduce the code that is
duplicated but that should still be good for CRing.

The goal is to run the exporter only once at the beginning of the
project, so fill the Shopify store with proper data. We were kind of
obligated to do that, since if we don't assign any variant to the
product when we create it, it assigns a dummy variant that is hard to
get rid of. I tried just to export the master variant when creating a
new product but it was really pain in the arse. This came out to be the
best solution I could come up with.

thisiscab committed:
WIP - Export bundled products in Shopify

We permute all the part possibilities as Shopify Variants and we export
that to Shopify.

This is really really rough, lot of refactoring left to do, but the
basic implementation is there.

thisiscab committed:
Refactor the bundled product export

This way it's easier to know what's going on and know who does what.
There still a bit of refactor to do, but this works for now.

There are still specs to need to be written, this is the next step.

thisiscab committed:
Add specs in order to have to most coverage

This doesn't test everything but this should give us enough coverage to
see if something is wrong.

All the major cases and the main classes are tested to ensure that
everything works as expected.

thisiscab committed:
Remove shopify task

We do not want to have task step-up at the moment. We are not at a point
where we know what to run when. Let's leave the choice to the caller.

thisiscab committed:
Add a set of validations to make sure that is coherent

There are some possibilities where bundled products have no option
values, this will support that case. We were always expecting data to be
received but sometimes there are nil data that gets propagated thus it
was halting the export.

thisiscab committed:
Change the find_by methods

I have detected a problem with the base_decorator where it would not
find the proper variant. I was always testing with a Shopify Shop that
was empty and now that we have some real data a problem came out.

We were always returning all the variants even though we were trying to
find one with a specific id and we first picking the first one from that
array.

Instead, I've decided to use the `find` method that they are already
using in their ShopifyAPI gem and catching the 404 error. When that
exception is caught we create a new instance of that object and return
it.

This also regenerate the cassettes, due to the fact that the
base_decorator has changed which indirectly make the requests change.

thisiscab committed:
Merge branch 'master' into feature/shopify/bundle-products-export

braidn commented on app/attributes/shopify/bundled_product_attributes.rb:
```diff
@@ -0,0 +1,48 @@
+      @bundle = bundle
+      @converter = product_converter
+      @part_attributes = part_attributes
+      @permuter = permuter
```
What's a permuter? The word is a little bizarre to me, just need a little context maybe? Thanks.

thisiscab replied:
It takes all the possible option_values of the bundle and it creates all the different permutation. 
Let's say we have the following case instance:
{ { BlueGloss, RedGloss, PinkGloss }
{ RedLip, PurpleLip, PinkLip } }

the outcome of this will be the permutation of every scenario, which would output an array with 9 possible permutation.

braidn replied:
Is permuter a word?

braidn commented on app/converters/shopify/part_converter.rb:
```diff
@@ -0,0 +1,31 @@
+      @permutation = permutation
+    end
+
+    def to_hash
```
Does it make sense to have this be a method that's called on a new object? Or is this more of a utility object that could be called with PartConverter.to_hash(part:, permutation:)? Maybe it's more of a style question than anything.

thisiscab replied:
Yeah I think you're right. It doesn't make much sense to have it as being a class. But at the same time I thought of ActiveRecord that converts a regular object `.to_hash` without any parameters and converts the object to an hash. 

I see your point but I am unsure if this would "confuse" people that you have to pass params to the `to_hash` method, but at the same time, it's confusing having to initialize it. IDK :)

braidn replied:
Good point, it might confuse folks.

braidn commented on app/attributes/shopify/bundled_product_attributes.rb:
```diff
@@ -0,0 +1,48 @@
+    end
+
+    def parts_attributes
+      attrs_var = { variants: [] }
```
Could/should this data object be instantiated in the initialize method?

braidn commented on app/converters/shopify/bundle_converter.rb:
```diff
@@ -0,0 +1,36 @@
+    attr_reader :bundle, :option_types
+
+    def all_option_types
+      options = { options: [] }
```
Along with this going under the same question of 'can it be pushed back into the initializer, are the outside brackets `{}` required? Doesn't this throw a rubocop error?

thisiscab replied:
{} are required, else ruby doesn't know how to compile this. 
Also Rubocop throws an error when I remove the {}

braidn replied:
Really? Weird. Ok. I would figure that Rubocop wasn't a fan of the explicit brackets and if it breaks, hey. I was super wrong.

braidn commented on app/exporters/shopify/product_exporter.rb:
```diff
@@ -1,48 +1,62 @@
     end
 
-    def perform
+    def save_product_on_shopify
```
I agree this reads better but it locks us into the concept of Shopify. I know, I know there is more use of it throughout the gem but we need to take an active move to maybe remove it. Since this is quite a large factory style class (sure it doesn't create other classes here but it does in Shopify), `perform` or `call` seem like a cleaner interface, no?

thisiscab replied:
Yes! I agree with keeping `perform`, let's do eit! :)

braidn commented on app/exporters/shopify/product_exporter.rb:
```diff
@@ -1,48 +1,62 @@
 
-    def product_attributes
-      product_converter.new(product: presented_product).to_hash
+    class AssociationSaver
```
The case for keeping this in this file is?

braidn commented on app/presenters/shopify/part_presenter.rb:
```diff
@@ -0,0 +1,30 @@
+      product.master
+    end
+
+    def sku_separator
```
Since we are interpreting the string above, this seems like a whole lot more characters to write over just adding in the backtick which is a single char?

thisiscab replied:
This allows for people to overwrite that string easily by decorating the method return value. 
I thought this was a nice to add, whatcha think?

braidn replied:
I mean it's nice, sure. But do we know people who use this will even use the bundle extension? Perhaps Glossier will be the only folks who use the gem with bundled products. It feels a little like premature extension but, it's totally fine.

braidn commented on app/updaters/shopify/product_updater.rb:
```diff
@@ -0,0 +1,43 @@
+      product_api.find_or_initialize_by_id(spree_product.pos_product_id)
+    end
+
+    # FIXME: refactor this
```
Should we have a `Refactor.md` file for things like this?

braidn commented on app/updaters/shopify/product_updater.rb:
```diff
@@ -0,0 +1,43 @@
+      shopify_product = find_shopify_product_for(spree_product)
+
+      if shopify_product.persisted?
+        shopify_product.update_attributes(product_attributes)
```
Could these two items be pushed down into private and this made a guard clause?

braidn commented on app/updaters/shopify/variant_updater.rb:
```diff
@@ -1,43 +1,41 @@
+      variant_api.find_or_initialize_by_id(spree_variant.pos_variant_id, params: { product_id: spree_variant.product.pos_product_id })
     end
 
+    # FIXME: refactor this
```
`Refactor.md` idea again.

braidn commented on spec/support/shared/spree_builders.rb:
```diff
@@ -1,10 +1,14 @@
                           created_at: build_date_time,
                           updated_at: build_date_time,
-                          available_on: build_date_time)
+                          available_on: build_date_time,
```
Breaks, breaks, breaks! This is a lot of coupled data. Is there a better way to stub boundaries without recreating the whole data object as a double?

thisiscab replied:
I would agree that this is getting out of hand quick. But this really helps in term of the speed of the unit test. This allow us to inject stuff easily. I am unsure how else to proceed, this was suggested by DW and I thought I would give it a whirl. 

Do you have suggestions?

braidn reviewed (commented):
Ok, this is a v1. There is way too much going on in this PR for any one person to catch it all. It's a start.

thisiscab committed:
Merge branch 'master' into feature/shopify/bundle-products-export

thisiscab committed:
Change the caller method name for some classes

As per the CR, it would read better to keep `perform` instead of
`save_product_on_shopify`. Also this would actively lock us out of
Shopify if we ever want to. And I personally feel like it reads better.

thisiscab committed:
Call the helper association object instead of a self defined one

This will help us change data at one place instead of having to change
it at multiple places.

thisiscab commented:
@braidn I have merged the latest master. The PR is now smaller than it was previously. 
I have answered some of your comments and I'd like to continue the conversation on some of those. 

I'll wait to hear from you! :)

braidn commented:
Commented

thisiscab commented:
Thank you very much sir! :)
